src/openjd/model/_range_expr.py

```diff
@@ -156,31 +165,58 @@ def _validate(self) -> None:
                 )
             prev_range = range_
 
+    @classmethod
+    def _pydantic_validate(cls, value: Any) -> Any:
+        if isinstance(value, IntRangeExpr):
+            return value
+        elif isinstance(value, str):
+            return IntRangeExpr.from_str(value)
+        elif isinstance(value, list):
+            return IntRangeExpr.from_list(value)
+        else:
+            raise ValueError("Value must be an integer range expression or a list of integers.")
+
+    @classmethod
+    def __get_pydantic_core_schema__(
+        cls, source_type: type[Any], handler: GetCoreSchemaHandler
+    ) -> core_schema.CoreSchema:
+        return core_schema.no_info_plain_validator_function(cls._pydantic_validate)
+
+    @classmethod
+    def __get_pydantic_json_schema__(
+        cls, core_schema: core_schema.CoreSchema, handler: GetJsonSchemaHandler
+    ) -> JsonSchemaValue:
+        return {"type": "string"}
+
 
-@total_ordering
 class IntRange(Sized):
-    """Inclusive on the start and end value"""
+    """A linear sequence of integers.
+
+    Both _start and _end are always included in the set of values, and _step is always positive."""
 
     _start: int
     _end: int
     _step: int
     _range: range
 
     def __init__(self, start: int, end: int, step: int = 1):
-        self._start = start
-        self._end = end
-        self._step = step
-
-        # makes the range inclusive on end value
-        offset = 0
-        if self._step > 0:
-            offset = 1
-        elif self._step < 0:
-            offset = -1
-
-        self._range = range(self._start, self._end + offset, self._step)
-
-        self._validate()
+        if step > 0:
+            if start > end:
+                raise ValueError("Range: a descending range must have a negative step")
+            self._range = range(start, end + 1, step)
```

Comment:
Should this be `range(start, end + step, step)` in case the step is not 1 but we still want to include the end value? Similar in the negative branch below. Or if this does work as expected, is there a test case that would verify the final value is included in a range with a different step size?

Reply:
Consider these cases:

```
>>> "1-5:3" # 1 to 5 (inclusive), stepping by 3. Values [1, 4]
'1-5:3'
>>> list(range(1,5 + 1,3)) # With + 1
[1, 4]
>>> list(range(1,5 + 3,3)) # With + step
[1, 4, 7]
```

It wasn't obvious whether this was being tested thoroughly, so I dropped in a few more test cases.

Reply:
Ah, I had a bad assumption that ranges could include floats. Looks like that's not true of OJD or Python's `range`.